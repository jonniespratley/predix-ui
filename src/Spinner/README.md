The `<Spinner/>` indicates an ongoing activity with an indefinite end, such as loading data. 

## Usage

```react
const {Spinner} = PxReact;
<Spinner size={100}/>
```

### Custom CSS example


```react
const {Spinner} = PxReact;
<div className='custom'>
  <style>{`
    .custom{
      --px-spinner-fill-color: green;
      --px-spinner-background-color: pink;
    }
    `}</style>
  <Spinner size={100}/>
</div>
```


### Styling
The following custom properties are available for styling:
```table
span: 6
rows:
  - Name: --px-spinner-fill-color
    Description: Fill color for the spinner
    Default: black
  - Name: --px-spinner-background-color
    Description: Background color for the spinner
    Default: gray
```
