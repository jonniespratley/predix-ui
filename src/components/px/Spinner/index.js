import React from 'react';
import PropTypes from 'prop-types';
import styled, { keyframes } from 'styled-components';

const SpinnerRotateAnimation = keyframes`
 to {
    stroke-dashoffset: 0;
  }
`;
/*
const SpinnerAnimation = keyframes`
  0% {
    -webkit-transform: rotate(0deg);
    transform: rotate(0deg);
  }
  100% {
    -webkit-transform: rotate(360deg);
    transform: rotate(360deg);
  }
`; */

const SpinnerCircle1 = styled.circle`
  fill: none;
  stroke: var(--px-spinner-background-color, gray);
  stroke-width: 5px;
`;

const SpinnerCircle2 = styled.circle`
  fill: none;
  stroke: var(--px-spinner-fill-color, black);
  stroke-width: 5px;
  stroke-dasharray: 283;
  stroke-dashoffset: 566;
  animation: ${SpinnerRotateAnimation} 1.5s infinite cubic-bezier(0.78, 0.13, 0.16, 0.87);
`;

const Component = ({ size, finished }) => (
  <div>
    {!finished &&
    <svg viewBox="0 0 100 100" height={size} width={size}>
      <SpinnerCircle1 cx="50" cy="50" r="45" />
      <SpinnerCircle2 cx="50" cy="50" r="45" transform="rotate(-45,50,50)" />
    </svg>}
  </div>
);

Component.defaultProps = {
  size: 100,
  finished: false
};

Component.propTypes = {
  size: PropTypes.number,
  finished: PropTypes.bool
};

Component.displayName = 'Spinner';

export default Component;
