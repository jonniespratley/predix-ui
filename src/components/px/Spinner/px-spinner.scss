$px-spinner-background-color: var(--px-spinner-background-color, gray);
$px-spinner-fill-color: var(--px-spinner-fill-color, black);

.px-spinner {
	.circle1 {
		fill        : none;
		stroke      : $px-spinner-background-color;
		stroke-width: 5px;
	}

	.circle2 {
		fill             : none;
		stroke           : $px-spinner-fill-color;
		stroke-width     : 5px;
		stroke-dasharray : 283;
		stroke-dashoffset: 566;
		animation        : pxSpinnerRotate 1.5s infinite cubic-bezier(.78,.13,.16,.87);
	}
}


// IE/Edge version (does not support SVG animation)

.spinner-container {
	font-size    : 10px;
	position     : relative;
	text-indent  : -9999em;
	border-top   : 5px solid $px-spinner-background-color;
	border-right : 5px solid $px-spinner-background-color;
	border-bottom: 5px solid $px-spinner-background-color;
	border-left  : 5px solid $px-spinner-fill-color;
	transform    : translateZ(0);
	animation    : pxSpinnerLoad8 1.1s infinite cubic-bezier(.78,.13,.16,.87);
}

.spinner-container,
.spinner-container:after {
	border-radius: 50%;
	width        : 10em;
	height       : 10em;
}
@keyframes pxSpinnerLoad8 {
	0% {
		-webkit-transform: rotate(0deg);
		transform        : rotate(0deg);
	}

	100% {
		-webkit-transform: rotate(360deg);
		transform        : rotate(360deg);
	}
}
@keyframes pxSpinnerRotate {
	to {
		stroke-dashoffset: 0;
	}
}
