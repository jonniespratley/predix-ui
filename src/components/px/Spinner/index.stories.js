import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, boolean, number } from '@storybook/addon-knobs/react';

import Spinner from './';


// Spinner
storiesOf('Components / Spinner', module)
  .addDecorator(withKnobs)
  .addWithJSX('default', () => (
    <Spinner
      finished={boolean('finished', false)}
      size={number('size', 125)}
    />
  ));
