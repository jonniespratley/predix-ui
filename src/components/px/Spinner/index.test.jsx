import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import Spinner from './';

describe('Spinner', () => {
  test('should render', () =>{
    const tree = renderer.create(<Spinner/>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('should render with value', () =>{
    const tree = renderer.create(<Spinner value={50}/>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
