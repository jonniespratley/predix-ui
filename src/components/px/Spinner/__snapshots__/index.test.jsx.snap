// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner should render 1`] = `
.c0 {
  fill: none;
  stroke: var(--px-spinner-background-color,gray);
  stroke-width: 5px;
}

.c1 {
  fill: none;
  stroke: var(--px-spinner-fill-color,black);
  stroke-width: 5px;
  stroke-dasharray: 283;
  stroke-dashoffset: 566;
  -webkit-animation: ffGLem 1.5s infinite cubic-bezier(0.78,0.13,0.16,0.87);
  animation: ffGLem 1.5s infinite cubic-bezier(0.78,0.13,0.16,0.87);
}

<div>
  <svg
    height={100}
    viewBox="0 0 100 100"
    width={100}
  >
    <circle
      className="c0"
      cx="50"
      cy="50"
      r="45"
    />
    <circle
      className="c1"
      cx="50"
      cy="50"
      r="45"
      transform="rotate(-45,50,50)"
    />
  </svg>
</div>
`;

exports[`Spinner should render with value 1`] = `
.c0 {
  fill: none;
  stroke: var(--px-spinner-background-color,gray);
  stroke-width: 5px;
}

.c1 {
  fill: none;
  stroke: var(--px-spinner-fill-color,black);
  stroke-width: 5px;
  stroke-dasharray: 283;
  stroke-dashoffset: 566;
  -webkit-animation: ffGLem 1.5s infinite cubic-bezier(0.78,0.13,0.16,0.87);
  animation: ffGLem 1.5s infinite cubic-bezier(0.78,0.13,0.16,0.87);
}

<div>
  <svg
    height={100}
    viewBox="0 0 100 100"
    width={100}
  >
    <circle
      className="c0"
      cx="50"
      cy="50"
      r="45"
    />
    <circle
      className="c1"
      cx="50"
      cy="50"
      r="45"
      transform="rotate(-45,50,50)"
    />
  </svg>
</div>
`;
